File: web/src/api/content.ts
import { request } from "./http";
import type { Article } from "./articles";

export type Series = {
  id: number;
  slug: string;
  title: string;
  description: string;
  cover_image_url: string | null;
  sort_order: number;
  article_count: number;
  created_at: string;
  updated_at: string;
};

export type SeriesDetail = Series & { articles: Article[] };

export type Note = {
  id: number;
  slug: string;
  content_markdown: string;
  tags: string[];
  external_url: string | null;
  published_at: string | null;
  created_at: string;
  updated_at: string;
};

export type NoteList = {
  items: Note[];
  total: number;
  page: number;
  page_size: number;
};

export function fetchSeries() {
  return request<{ items: Series[]; total: number }>({ url: "/series", method: "GET" });
}

export function fetchSeriesDetail(slug: string) {
  return request<SeriesDetail>({ url: `/series/${encodeURIComponent(slug)}`, method: "GET" });
}

export function fetchNotes(params: Record<string, string | number | undefined> = {}) {
  return request<NoteList>({ url: "/notes", method: "GET", params });
}

export function fetchNote(slug: string) {
  return request<Note>({ url: `/notes/${encodeURIComponent(slug)}`, method: "GET" });
}

File: backend/app/models/content.py
from datetime import datetime

from sqlalchemy import DateTime, Integer, JSON, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base


class Series(Base):
    __tablename__ = "series"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    slug: Mapped[str] = mapped_column(String(160), unique=True, index=True)
    title: Mapped[str] = mapped_column(String(200))
    description: Mapped[str] = mapped_column(Text, default="")
    cover_image_url: Mapped[str | None] = mapped_column(String(2048), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=func.now(),
        onupdate=func.now(),
        server_default=func.now(),
    )


class Note(Base):
    __tablename__ = "notes"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    slug: Mapped[str] = mapped_column(String(160), unique=True, index=True)
    content_markdown: Mapped[str] = mapped_column(Text)
    tags: Mapped[list[str]] = mapped_column(JSON, default=list)
    external_url: Mapped[str | None] = mapped_column(String(2048), nullable=True)
    published_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=func.now(),
        onupdate=func.now(),
        server_default=func.now(),
    )
